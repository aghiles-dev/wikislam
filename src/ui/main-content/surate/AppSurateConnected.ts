import { bindActionCreators, Dispatch } from 'redux'
import { connect } from 'redux-vue'
import { getAllSurates, getCurrentSurate, isSuratesListBeingFetched } from '../../../config/store/rootSelectors'
import { Wording, WordingConstants } from '../../../app/shared/wordingConstants'
import { AppState } from '../../../config/store/rootState'
import AppSurates from './AppSurate.vue'
import { surateActions } from '../../../app/surate/usecases/surate.actions'
import { AppRouter } from '../../../config/router/router'
import { Verse } from '../../../app/surate/domain/Verse'
import { SurateState } from '../../../app/surates-list/domain/SurateState'

export type AppSurateProps = {
  surateId: number
  surateDetails?: SurateState
  verses: Verse[]
  isLoading: boolean
  wording: Wording
}

export type AppSurateActions = {
  fetchSurateById: (surateId: number) => void
  resetCurrentSurate: () => void
}

const mapStateToProps = (state: AppState): { state: AppSurateProps } => ({
  state: {
    surateId: +AppRouter.currentRoute.params.surateId,
    surateDetails: getAllSurates(state).find(surate => surate.id === +AppRouter.currentRoute.params.surateId),
    verses: getCurrentSurate(state).verses,
    isLoading: getCurrentSurate(state).isLoading,
    wording: WordingConstants
  }
})

const mapDispatchToProps = (dispatch: Dispatch): { actions: AppSurateActions } => ({
  actions: bindActionCreators(
    {
      fetchSurateById: surateActions.fetchSurateById,
      resetCurrentSurate: surateActions.resetCurrentSurate
    },
    dispatch
  )
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AppSurates)
