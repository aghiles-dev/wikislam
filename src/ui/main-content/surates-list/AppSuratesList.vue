<template>
  <div>
    <progress v-if="state.isLoading" class="progress is-small is-primary" max="100"/>

    <div v-else>
      <div class="columns is-centered"
           v-for="surate in state.surates"
           :key="surate.id">
        <div class="column card is-10-mobile is-8-tablet">
          <p class="is-size-4 has-text-centered has-cursor-pointer"
             @click="goToSurate(surate.id)">
            <span>{{surate.surateNumber}}. {{surate.name}} ({{surate.numberOfVerses}})({{surate.arabicName}})</span><br>
            <span class="is-size-5">{{surate.frenchName}}</span>
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
  import Vue from 'vue'
  import { Component, Prop } from 'vue-property-decorator'
  import { AppSuratesListActions, AppSuratesListProps } from './AppSuratesListConnected'
  import { AppRouteName } from '../../../config/router/router'

  @Component({})
  export default class AppSuratesList extends Vue {
    @Prop() state: AppSuratesListProps
    @Prop() actions: AppSuratesListActions

    mounted () {
      this.actions.fetchSuratesList()
    }

    goToSurate (surateId) {
      this.$router.push({ name: AppRouteName.SURATE_DETAILS, params: { surateId } })
    }
  }
</script>

<style lang="scss" scoped>

</style>
