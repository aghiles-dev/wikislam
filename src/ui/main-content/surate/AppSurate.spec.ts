import { shallowMount } from '@vue/test-utils'
import AppSuratesList from './AppSurate.vue'
import { AppSurateActions, AppSurateProps } from './AppSurateConnected'

describe('Component | AppSurates', () => {
  let state: AppSurateProps
  let actions: AppSurateActions

  beforeEach(() => {
    state = {
      surateId: 1,
      verses: [
        {
          id: 1,
          verseNumber: 1,
          chapterId: 2,
          verseKey: '2:1',
          arabicText: 'fake arabic text',
          frenchText: 'fake french text',
          juzNumber: 1,
          hizbNumber: 1,
          sajdah: false
        }
      ],
      isLoading: false,
      wording: {}
    }

    actions = {
      fetchSurateById: jest.fn()
    }
  })

  describe('When the surate is being fetched', () => {
    it('should match App snapshot', () => {
      // Given
      const wrapper = shallowMount(AppSuratesList, {
        propsData: {
          state: { ...state, isLoading: true },
          actions
        }
      })

      // Then
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('When the surate has already been fetched', () => {
    it('should match snapshot', () => {
      // Given
      const wrapper = shallowMount(AppSuratesList, {
        propsData: { state, actions }
      })

      // Then
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('on init', () => {
    it('fetches all the surates', () => {
      // When
      shallowMount(AppSuratesList, {
        propsData: { state, actions }
      })

      // Then
      expect(actions.fetchSurateById).toHaveBeenCalledTimes(state.surateId)
    })
  })

})
