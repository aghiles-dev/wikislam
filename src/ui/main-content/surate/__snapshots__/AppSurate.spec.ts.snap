// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | AppSurates When the surate has already been fetched should match snapshot 1`] = `
<div>
  <div>
    
    in
  
  </div>
</div>
`;

exports[`Component | AppSurates When the surate is being fetched should match App snapshot 1`] = `
<div>
  <progress
    class="progress is-small is-primary"
    max="100"
  />
</div>
`;
